src/fsharp/ParseHelpers.fs

```diff
@@ -177,6 +177,48 @@ let rec LexerIfdefEval (lookup: string -> bool) = function
     | IfdefNot e        -> not (LexerIfdefEval lookup e)
     | IfdefId id        -> lookup id
 
+/// Ifdef F# lexer/parser state, held in the BufferLocalStore for the lexer.
+/// Used to capture #if, #else and #endif as syntax trivia.
+module LexbufIfdefStore =
+    // The key into the BufferLocalStore used to hold the compiler directives
+    let private ifDefKey = "Ifdef"
+
+    let private getStore (lexbuf: Lexbuf): ResizeArray<ConditionalDirectiveTrivia> =
+        match lexbuf.BufferLocalStore.TryGetValue ifDefKey with
+        | true, store -> store
+        | _ ->
+            let store = box (ResizeArray<ConditionalDirectiveTrivia>())
+            lexbuf.BufferLocalStore.[ifDefKey] <- store
+            store
+        |> unbox<ResizeArray<ConditionalDirectiveTrivia>>
+
+    let SaveIfHash (lexbuf: Lexbuf, expr: LexerIfdefExpression, range: range) =
+        let store = getStore lexbuf
+
+        let expr =
+            let rec visit (expr: LexerIfdefExpression) : IfDirectiveExpression =
+                match expr with
+                | LexerIfdefExpression.IfdefAnd(l,r) -> IfDirectiveExpression.IfdefAnd(visit l, visit r)
+                | LexerIfdefExpression.IfdefOr(l, r) -> IfDirectiveExpression.IfdefOr(visit l, visit r)
+                | LexerIfdefExpression.IfdefNot e -> IfDirectiveExpression.IfdefNot(visit e)
+                | LexerIfdefExpression.IfdefId id -> IfDirectiveExpression.IfdefId id
+            
+            visit expr
+
+        store.Add(ConditionalDirectiveTrivia.IfDirectiveTrivia(expr, range))
```

Comment:
I'm not sure if the condition for the ifdef expression is correctly named trivia.

Reply:
That was inspired by https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.conditionaldirectivetriviasyntax?view=roslyn-dotnet-4.1.0 per [Don's request](https://github.com/dotnet/fsharp/issues/12815).
I'm all ears if you want to go with another name.

Reply:
naming is the toughest problem in computer science.

Reply:
It's interesting that roslyn has properties Condition and ConditionValue, which sort of indicates the Trivia means (with Trivia).  So the name you proposed matches the prior art.